File: internal/cmdutil/paginate_test.go
package cmdutil

import (
	"net/url"
	"testing"
)

func TestNextCursor(t *testing.T) {
	page := map[string]any{
		"next":     "https://assess.praxicraft.com/api/v1/public/assessments/?cursor=abc123&page_size=20",
		"previous": nil,
		"results":  []any{},
	}
	if got := NextCursor(page); got != "abc123" {
		t.Fatalf("NextCursor=%q want abc123", got)
	}
	if NextCursor(map[string]any{"next": nil}) != "" {
		t.Fatal("expected empty cursor")
	}
}

func TestNextPage(t *testing.T) {
	page := map[string]any{
		"next": "https://example.com/results/?page=2",
	}
	if got := NextPage(page); got != "2" {
		t.Fatalf("NextPage=%q want 2", got)
	}
}

func TestFetchAllMergesPages(t *testing.T) {
	calls := 0
	fetch := func(q url.Values) (any, error) {
		calls++
		switch calls {
		case 1:
			if q.Get("page_size") != "100" {
				t.Fatalf("expected default page_size=100, got %q", q.Get("page_size"))
			}
			return map[string]any{
				"next": "https://x/?cursor=c2",
				"results": []any{
					map[string]any{"slug": "a"},
					map[string]any{"slug": "b"},
				},
			}, nil
		case 2:
			if q.Get("cursor") != "c2" {
				t.Fatalf("expected cursor=c2, got %q", q.Get("cursor"))
			}
			return map[string]any{
				"next": nil,
				"results": []any{
					map[string]any{"slug": "c"},
				},
			}, nil
		default:
			t.Fatal("unexpected extra fetch")
			return nil, nil
		}
	}
	out, err := FetchAll(fetch, nil)
	if err != nil {
		t.Fatal(err)
	}
	rows := ResultsRows(out)
	if len(rows) != 3 {
		t.Fatalf("got %d rows want 3", len(rows))
	}
	m := out.(map[string]any)
	if m["next"] != nil {
		t.Fatalf("merged next should be nil, got %v", m["next"])
	}
}

func TestAdvanceQuery(t *testing.T) {
	q := url.Values{}
	ok := AdvanceQuery(q, map[string]any{
		"next": "https://x/?cursor=zz",
	})
	if !ok || q.Get("cursor") != "zz" {
		t.Fatalf("AdvanceQuery cursor failed: ok=%v q=%v", ok, q)
	}
}

File: tests/paginate.test.ts
import { describe, expect, test } from "bun:test";
import {
  advanceQuery,
  fetchAllPages,
  nextCursor,
  nextPageNumber,
  resultsRows,
  stripAllFlag,
} from "../src/utils/paginate";

describe("paginate", () => {
  test("nextCursor extracts cursor from next URL", () => {
    expect(
      nextCursor({
        next: "https://assess.praxicraft.com/api/v1/public/assessments/?cursor=abc123&page_size=20",
        results: [],
      }),
    ).toBe("abc123");
    expect(nextCursor({ next: null })).toBe("");
  });

  test("nextPageNumber extracts page from next URL", () => {
    expect(nextPageNumber({ next: "https://example.com/results/?page=2" })).toBe("2");
  });

  test("advanceQuery sets cursor query param", () => {
    const query: Record<string, string> = {};
    expect(advanceQuery(query, { next: "https://x/?cursor=zz" })).toBe(true);
    expect(query.cursor).toBe("zz");
  });

  test("fetchAllPages merges results and clears next", async () => {
    let calls = 0;
    const out = await fetchAllPages(async (query) => {
      calls++;
      if (calls === 1) {
        expect(query.page_size).toBe("100");
        return {
          next: "https://x/?cursor=c2",
          results: [{ slug: "a" }, { slug: "b" }],
        };
      }
      expect(query.cursor).toBe("c2");
      return {
        next: null,
        results: [{ slug: "c" }],
      };
    });

    expect(resultsRows(out)).toHaveLength(3);
    expect((out as Record<string, unknown>).next).toBeNull();
    expect(calls).toBe(2);
  });

  test("stripAllFlag removes --all", () => {
    expect(stripAllFlag(["foo", "--all"])).toEqual({ args: ["foo"], all: true });
    expect(stripAllFlag(["foo"])).toEqual({ args: ["foo"], all: false });
  });
});
